spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DistriParameterSynchronizer.scala

```diff
@@ -0,0 +1,429 @@
+
+  private val clearPoolSize: Int = System.getProperty("bigdl.ParameterSynchronier." +
+    "clearPoolSize", "1").toInt
+
+  private val workerPoolSize: Int = System.getProperty("bigdl.ParameterSynchronier" +
+    ".syncPoolSize", "4").toInt
+
+  private val syncPoolSize: Int = Math.max(System.getProperty("bigdl.ParameterSynchronier" +
+    ".computePoolSize",
+    (Runtime.getRuntime().availableProcessors() / 2).toString).toInt, 2)
+
+  private val fetchCompletionPoolSize: Int = System.getProperty("bigdl.ParameterSynchronier" +
+    ".fetchCompletionPoolSize", "1").toInt
+
+  private val asyncTaskWaitingQueue : PriorityBlockingQueue[AsyncFutureTask] =
+    new PriorityBlockingQueue[AsyncFutureTask](taskSize)
+
+  private val blockFetchRequestQueue: PriorityBlockingQueue[BlockFetchRequest] =
+    new PriorityBlockingQueue[BlockFetchRequest](taskSize)
+
+  // thread pool to remove expired blocks
+  private lazy val clearPool: ExecutorService =
+    Executors.newFixedThreadPool(clearPoolSize, new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // main thread pool to do put-get-aggregate
+  private val workerPool: ExecutorService =
+    Executors.newFixedThreadPool(workerPoolSize, new ThreadFactory {
+    override def newThread(r: Runnable): Thread = {
+      val t = Executors.defaultThreadFactory().newThread(r)
+      t.setDaemon(true)
+      t
+    }
+  })
+
+  // long running thread to fetch the request
+  workerPool.submit(new Runnable {
+    override def run(): Unit = {
+      val asyncFutureTask = asyncTaskWaitingQueue.take
+      workerPool.submit(asyncFutureTask.task)
+    }
+  })
+
+  // thread pool for put and aggregate
+  private lazy val syncPool: ExecutorService = Executors.newFixedThreadPool(syncPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // thread pool for fetching blocks
+  private lazy val fetchPool: ExecutorService = Executors.newFixedThreadPool(syncPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // thread pool to update sow on fetching completion
+  private val fetchCompletionPool: ExecutorService = Executors.
+    newFixedThreadPool(fetchCompletionPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  (0 until syncPoolSize).foreach(th => {
+    fetchPool.submit(new Runnable {
+      override def run(): Unit = {
+        while (!shutdown) {
+          val fetchRequest = blockFetchRequestQueue.take
```

Comment:
even you set shutdown == true, if the thread is blocked at the `take`, the thread will not shutdown, right?

Reply:
add a graceful mechanism to exit the thread after shutdown